src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1277,28 +1252,48 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
 
         // Only render attributes if client-side validation is enabled, and then only if we've
         // never rendered validation for a field with this name in this form.
-        protected virtual IDictionary<string, object> GetValidationAttributes(
+        public virtual void AddValidationAttributes(
```

Comment:
Wondering about exposing this in `IHtmlGenerator`. Thoughts?

Going the other way, should it be `protected virtual` like most of the helpers in this class?

In any case, should have doc comments.